bitcoin/src/blockdata/witness.rs

```diff
@@ -384,24 +384,19 @@ impl Witness {
     /// [Script::is_p2tr](crate::blockdata::script::Script::is_p2tr) to
     /// check whether this is actually a Taproot witness.
     pub fn tapscript(&self) -> Option<&Script> {
-        let len = self.len();
-        self.last()
-            .map(|last_elem| {
-                // From BIP341:
-                // If there are at least two witness elements, and the first byte of
-                // the last element is 0x50, this last element is called annex a
-                // and is removed from the witness stack.
-                if len >= 2 && last_elem.first() == Some(&TAPROOT_ANNEX_PREFIX) {
-                    // account for the extra item removed from the end
-                    3
-                } else {
-                    // otherwise script is 2nd from last
-                    2
-                }
-            })
-            .filter(|&script_pos_from_last| len >= script_pos_from_last)
-            .and_then(|script_pos_from_last| self.nth(len - script_pos_from_last))
-            .map(Script::from_bytes)
+        self.last().and_then(|last| {
+            // From BIP341:
+            // If there are at least two witness elements, and the first byte of
+            // the last element is 0x50, this last element is called annex a
+            // and is removed from the witness stack.
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
```

Comment:
Why does it compare to 3 if it says 2? And why didn't tests trigger? A job for mutagen?

Reply:
I think the comment is just confusing. Basically, if there is only one element then it's a signature. If there are 2 or more, then the final element is a control block, unless the final element starts with `0x50`, in which case it's an annex and the second-to-last element is the control block.

Maybe the logic can be rearranged to make this clearer.

Reply:
The comment was taken from the other method that does the same thing. So yeah there can only be an annex if there are at least 3 elements (optional script witness elements (can be 0), witness script, control block and then potentially annex). I don't know how it works exactly that the control block can't start with 0x50 but I guess Sipa thought about that ;) The first byte of the control block is a flag byte anyway IIRC, so I think the 0x50 is just not a valid value for that byte.